src/core/util/latent_see.cc

```diff
@@ -0,0 +1,114 @@
+// Copyright 2024 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/core/util/latent_see.h"
+
+#include <chrono>
+#include <cstdint>
+#include <ratio>
+
+#include "absl/strings/str_cat.h"
+#include "absl/types/optional.h"
+
+#ifdef GRPC_ENABLE_LATENT_SEE
+namespace grpc_core {
+namespace latent_see {
+
+thread_local std::vector<Log::Event> Log::thread_events_;
+thread_local uint64_t Log::thread_id_ = Log::Get().next_thread_id_.fetch_add(1);
+std::atomic<uint64_t> Flow::next_flow_id_{1};
+
+std::string Log::GenerateJson() {
+  std::vector<RecordedEvent> events;
```

Comment:
I was thinking there should be a final FlushThreadLog before generating JSON, so that unmatched events would be recorded (flows & parent scopes that were never destroyed, which could help debugging). But since the flush is thread-local, this could be challenging. Is the performance hit significant without thread-local event collections? The flush would happen ~twice as often as it does now.

Reply:
it's enough that we don't want it (will definitely overly skew results)

mostly we're going to use this with some kind of benchmark anyway (because the interesting cases are at the tail!) and so there's going to be thousands of events... we just don't care about the end